### ISSUE TITLE
Add `.toBeApplicable` matcher

### ISSUE THREAD

**Issue by mrazauskas** [tstyche/tstyche#97]
References:
- https://github.com/tsdjs/tsd/issues/206
- https://github.com/microsoft/TypeScript/pull/50820
- [Decorators in Typescript 5.0](https://devblogs.microsoft.com/typescript/announcing-typescript-5-0/#decorators)
- [JavaScript metaprogramming with the 2022-03 decorators API](https://2ality.com/2022/10/javascript-decorators.html)

@tbroyer Thanks for starting the discussion on decorator applicability testing. If you have a moment, could I ask for your feedback on the following thoughts? 

First of all, if it sounds interesting, you can simply migrate the type test to TSTyche and it will work as expected:

```ts
import { expect } from "tstyche";

class Base extends HTMLElement {
  foo() {}
}

const classDec = <T extends new (...args: any[]) => Base>(
  value: T,
  _context: ClassDecoratorContext<T>
) => {
  return class DecoratedClass extends value {};
};

(
  @classDec
  class extends Base {}
);

expect(
  (
    @classDec
    class {}
  )
).type.toRaiseError();

expect(
  (
    @classDec
    class extends HTMLElement {}
  )
).type.toRaiseError("Unable to resolve signature of class decorator");

expect(() => {
  @classDec
  abstract class Test extends Base {}
}).type.toRaiseError(1238);
```

If used without arguments, the `.toRaiseError()` matcher will match any and all semantic errors found in the expression. If substring(s) or error code(s) are provided, only those errors will be matched. To learn more, see documentation: https://tstyche.org/reference/expect-api#to-raise-error

---

All works, but the `try..catch` pattern feels somewhat strange. The thing is that TypeScript user does not expect errors while writing code. I mean, we are using TypeScript to get feedback in a form of errors with some guidance how to fix those. As the result, users expect to have working code, but not errors.

I have a feeling that testing tools should follow this path. Type tests should demonstrate how APIs can be used, instead of what errors are raised.

Hence `.not.toHaveProperty()` matcher was added recently to TSTyche. `.not.toBeCallableWith()` and `.not.toBeConstructableWith()` are coming soon (see #52). These are complicated. Some progress is done in private branches, not everything is visible in the PR.

---

What if the applicability of a decorator could be tested with a dedicated matcher? On surface it sounds easier to implement than `.toBeCallableWith()`.

How this new matcher should it named? Perhaps simply: `.toBeApplicableTo()`. Description: "Checks if the source decorator function can be applied to the target class or class member."

The above test becomes:

```ts
import { expect } from "tstyche";

class Base extends HTMLElement {
  foo() {}
}

const classDec = <T extends new (...args: any[]) => Base>(
  value: T,
  _context: ClassDecoratorContext<T>
) => {
  return class DecoratedClass extends value {};
};

abstract class Test extends Base {}

expect(classDec).type.toBeApplicableTo(class extends Base {});

expect(classDec).type.not.toBeApplicableTo(class {});
expect(classDec).type.not.toBeApplicableTo(class extends HTMLElement {});
expect(classDec).type.not.toBeApplicableTo(Test);
```

**Comment by tbroyer** [comment]
I didn't know about TSTyche so thanks for the ping (and knowing that it works with TSTyche makes me hopeful it can be made to work with tsd too).

Regarding your `toBeApplicableTo` proposal, while it works great for class decorators, how about class *member* decorators? I don't think you can pass a getter or method as an argument, and how about auto-accessors? But maybe *just* passing the member name as a second argument? (but then, how about private members? would that work? imagine a decorator declared as `function dec (value: undefined, context: ClassFieldDecoratorContext<Base, any> & { private: true })`; how about static and non-static members with the same name?)

**Comment by mrazauskas** [comment]
Thanks for good questions. To be honest I have zero experience with decorators, so implementing a matcher is a chance to learn something new.

As you propose, it sounds good to take second optional argument `key: string | number | symbol`:

```ts
expect(bound).type.toBeApplicableTo(Person, "greet");
expect(bound).type.toBeApplicableTo(Person, "#sleep");
```

I don’t see how else a private field like `#x` could be reached from outside.

---

Regarding tsd. I would say that _de facto_ it is an abandoned project, unfortunately. Owners / maintainers do almost nothing with it for years. If you look at last commits, these are mostly new error codes added by contributors. The maintainers merge those without asking to included tests. And so on. Maybe some magic will happen with https://github.com/tsdjs/tsd/issues/196, but it also looks rather forgotten.

**Comment by mrazauskas** [comment]
Ah.. static and non-static members with the same name is great edge case. Looks simple on surface:

```ts
class Foo {
  static bar() {
    return "xyz";
  }

  bar() {
    return 123;
  }
}

expect(bound).type.toBeApplicableTo(Foo.bar); // static method
expect(bound).type.toBeApplicableTo(Foo, "bar"); // same as '.toBeApplicableTo(Foo.prototype.bar)'.
```

But it is allowed to have `static #bar` or `private static bar` as well. If same name is the only problem, a hint could be used to distinguish between instance and static fields:

```ts
expect(bound).type.toBeApplicableTo(Foo, "bar"); // the default hint is "instance"
expect(bound).type.toBeApplicableTo(Foo, "bar", "static");
```

**Comment by mrazauskas** [comment]
What if this matcher would be a decorator?

```ts
import { expect } from "tstyche";

class Person {
  #name: string;
  constructor(name: string) {
    this.#name = name;
  }

  @expect.not.toBeApplicable(unbound) // <-- here it is
  greet() {
    console.log(`Hello, my name is ${this.#name}.`);
  }
}

const p = new Person("Ron");
p.greet();
```

**Comment by tbroyer** [comment]
Fwiw, I'm trying out TSTyche in a new project (if only because tsd can only test one .d.ts file at a time) and that `@expect().type.toBeApplicable()` would be a good addition!

I'm currently testing using `expect(class { … }).type.not.toRaiseError())` which means I either have a single `expect` and class for many members, or the tests are really verbose (I went with the former).

<details>
<summary>Comparing syntaxes</summary>

Currently:
```ts
expect(
  class {
    // ~80 properties/setters, testing ~40 decorators to be applicable

  @decorator1 accessor strProp: string;
  }
).type.not.toRaiseError();

expect(
  class {
    // one property, one decorator, one error
   @decorator2 accessor strProp: string;
  }
).type.toRaiseError(1240, 1270);
```

vs. with `@expect().type.toBeApplicable()`:

```ts
class {
  // ~80 properties/setters, testing ~40 decorators .toBeApplicable

  // also testing .not.toBeApplicable, possibly using many @expect on the same members

  @expect(decorator1).type.toBeApplicable
  @expect(decorator2).type.not.toBeApplicable
  accessor strProp: string;
}
```

or possibly:
```js
test("decorator1", () => {
  (
    class {
      // testing one decorator, both .toBeApplicable and .not.toBeApplicable
    }
  )
});

test("decorator2", () => {
  (
    class {
      // testing one decorator, both .toBeApplicable and .not.toBeApplicable
    }
  )
});

// …
```

</details>

**Comment by mrazauskas** [comment]
Indeed, with `@expect().type.toBeApplicable` all is much more clear.

After releasing TSTyche 3.0 recently, next I plan to work on [Ability matchers](https://github.com/tstyche/tstyche/discussions/224). Adding priority label for `.toBeApplicable`.

**Comment by mrazauskas** [comment]
@tbroyer Just to let you know, the latest beta has shipped with `.toBeApplicable`. Give it a try please.

There is one limitation: it can only handle types of expression, but not types. So, this works `@(expect(someDecorator).type.toBeApplicable)`, but this does not work yet `@(expect<SomeDecorator>().type.toBeApplicable)`.

The implementation is somewhat similar to [`ts-blank-space`](https://github.com/bloomberg/ts-blank-space). In a virtual file an assertion like `@(expect(someDecorator).type.toBeApplicable)` is turned into `@(       someDecorator                     )`. Any errors are mapped to the real file. In this case there is nothing much to map.

TSTyche prints out TS diagnostics, if any. Otherwise considers the function applicable. And `.not` inverts the logic. That’s it (;

**Comment by mrazauskas** [comment]
@tbroyer Lovely to see how you employed `.toBeApplicable` in [`es-parameters-decorator`](https://github.com/tbroyer/es-parameters-decorator). Let me know if something could be improved.

I think it is time to close this issue.

**Comment by tbroyer** [comment]
Yes, thank you for that!
